clouddrift/kinematics.py

```diff
@@ -3,9 +3,256 @@
+        np.argmin(np.abs(cor_freq[i] - radian_frequency / time_step))
+        for i in range(data_length)
+    ]
+
+    # get the transform at the inertial and "anti-inertial" frequencies
+    # extract the values of wp and wn at the calculated index as a function of time
+    # positive is anticyclonic (inertial) in the southern hemisphere
+    # negative is anticyclonic (inertial) in the northern hemisphere
+    wp = wp[frequency_bins, np.arange(0, data_length)]
+    wn = wn[frequency_bins, np.arange(0, data_length)]
+
+    # index of northen latitude points and index of
+    # southern latitude points
+    north = latitude >= 0
+
+    # initialize the zonal and meridional components of inertial displacements
+    wxhat = np.zeros_like(latitude, dtype=np.complex64)
+    wyhat = np.zeros_like(latitude, dtype=np.complex64)
+    # equations are x+ = 0.5*(z+ + z-) and y+ = -0.5*1j*(z+ - z-)
+    if any(north):
+        wxhat[north] = wn[north] / np.sqrt(2)
+        wyhat[north] = 1j * wn[north] / np.sqrt(2)
+    if any(~north):
+        wxhat[~north] = wp[~north] / np.sqrt(2)
+        wyhat[~north] = -1j * wp[~north] / np.sqrt(2)
+
+    # inertial displacement in meters
+    xhat = np.real(wxhat)
+    yhat = np.real(wyhat)
+
+    return xhat, yhat
+
+
+def residual_positions_from_displacements(
+    longitude: Union[float, np.ndarray],
+    latitude: Union[float, np.ndarray],
+    x: Union[float, np.ndarray],
+    y: Union[float, np.ndarray],
+) -> Union[Tuple[float], Tuple[np.ndarray]]:
+    """
+    Return residual longitudes and latitudes along a trajectory on the spherical Earth
+    after correcting for zonal and meridional displacements x and y in meters.
+
+    This is applicable as an example when one seeks to correct a trajectory for
+    horizontal oscillations due to inertial motions, tides, etc.
+
+    Parameters
+    ----------
+    longitude : float or np.ndarray
+        Longitude in degrees.
+    latitude : float or np.ndarray
+        Latitude in degrees.
+    x : float or np.ndarray
+        Zonal displacement in meters.
+    y : float or np.ndarray
+        Meridional displacement in meters.
+
+    Returns
+    ------
+    residual_longitude : float or np.ndarray
+        Residual longitude after correcting for zonal displacement, in degrees.
+    residual_latitude : float or np.ndarray
+        Residual latitude after correcting for meridional displacement, in degrees.
+
+    Examples
+    --------
+    Obtain the new geographical position for a displacement of 1/360-th of the
+    circumference of the Earth from original position (longitude,latitude) = (1,0):
+    >>> from clouddrift.sphere import EARTH_RADIUS_METERS
+    >>> residual_positions_from_displacements(1,0,2 * np.pi * EARTH_RADIUS_METERS / 360,0)
+    (0.0, 0.0)
+
+    Raises
+    ------
+    ValueError
+        If longitude, latitude, x, and y do not have the same shape.
+
+    """
+
+    if not longitude.shape == latitude.shape == x.shape == y.shape:
```

Comment:
```suggestion
    if (type(longitude) == np.ndarray or type(latitude) == np.ndarray or type(x) == np.ndarray or type(y) == np.ndarray) and not longitude.shape == latitude.shape == x.shape == y.shape:
```

this is ugly, but would work if all inputs are floats. Otherwise, it validates that they all have the same `shape`.

Reply:
Let's just let NumPy catch this:

```python
In [1]: from clouddrift.kinematics import residual_positions_from_displacements

In [2]: residual_positions_from_displacements(np.arange(10), np.arange(5), 10, 10)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 1
----> 1 residual_positions_from_displacements(np.arange(10), np.arange(5), 10, 10)

File ~/Work/clouddrift/clouddrift_selipot/clouddrift/kinematics.py:241, in residual_positions_from_displacements(longitude, latitude, x, y)
    235 longitudehat = (
    236     180 / np.pi * x / (EARTH_RADIUS_METERS * np.cos(np.radians(latitude)))
    237 )
    239 residual_latitude = latitude - latitudehat
    240 residual_longitude = recast_lon360(
--> 241     np.degrees(np.angle(np.exp(1j * np.radians(longitude - longitudehat))))
    242 )
    244 return residual_longitude, residual_latitude

ValueError: operands could not be broadcast together with shapes (10,) (5,) 
```

(the above is from the latest code without explicit raise)